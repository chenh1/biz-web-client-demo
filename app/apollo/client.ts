// import { ApolloLink, HttpLink } from "@apollo/client";
// import { onError } from "@apollo/client/link/error";

// import {
//   NextSSRInMemoryCache,
//   NextSSRApolloClient,
// } from "@apollo/experimental-nextjs-app-support/ssr";
// import { NextSSRInMemoryCache } from "@apollo/experimental-nextjs-app-support/ssr";
// import { registerApolloClient } from "@apollo/experimental-nextjs-app-support";}

// const errorLink = onError(({ graphQLErrors, networkError }) => {
//   if (graphQLErrors)
//     graphQLErrors.forEach(({ message, locations, path, ...rest }) => {
//       Object.keys(rest)?.map(key => console.log(key, rest[key]))
//       console.log(
//         `[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`
//       )
//     });
//   if (networkError) console.log(`[Network error]: ${networkError}`);
// });

// const httpLink = new HttpLink({
//   uri: `${process.env.API_URL || "http://127.0.0.1:1337"}/graphql`,
// })

// export const { getClient } = registerApolloClient(() => {
//   return new NextSSRApolloClient({
//     cache: new NextSSRInMemoryCache(),
//     link: ApolloLink.from([errorLink, httpLink]),
//   });
// });

import { HttpLink } from "@apollo/client";
import {
  registerApolloClient,
  ApolloClient,
  InMemoryCache,
} from "@apollo/experimental-nextjs-app-support";

export const { getClient, query, PreloadQuery } = registerApolloClient(() => {
  return new ApolloClient({
    cache: new InMemoryCache(),
    link: new HttpLink({
      // this needs to be an absolute url, as relative urls cannot be used in SSR
      uri: `${process.env.API_URL || "http://127.0.0.1:1337"}/graphql`,
      // you can disable result caching here if you want to
      // (this does not work if you are rendering your page with `export const dynamic = "force-static"`)
      // fetchOptions: { cache: "no-store" },
    }),
  });
});