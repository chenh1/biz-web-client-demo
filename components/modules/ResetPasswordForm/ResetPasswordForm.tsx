"use client";

import type { UserFormSubmitValues } from "@/components/constants/types";
import React, { useState } from "react"
import { useMutation } from "@apollo/client";
import { Box } from "../../containers/Box/Box";
import { Button } from "../../core/Button/Button";
import { Paragraph } from "../../core/Paragraph/Paragraph";
import { useRouter, useSearchParams } from "next/navigation";
import { InputText } from "../../fields/InputText/InputText";
import { resetPasswordMutation } from "../../../graphql/mutations/resetPassword";
import { Checkbox } from "../../fields/Checkbox/Checkbox";

export const ResetPasswordForm: React.FC = () => {
  const params = useSearchParams();
  const router = useRouter();
  const [ formError, setFormError ] = useState<string>("")
  const [ showPassword, setShowPassword ] = useState<boolean>(false);
  const code = params?.get("code");

  if (!code) {
    router.replace('/')
  }

  const [ resetPassword, { data } ] = useMutation(resetPasswordMutation);

  return (
    <form onSubmit={(e: UserFormSubmitValues) => {
      e.preventDefault()
      const password = e.target.password.value
      const passwordConfirmation = e.target.passwordConfirmation.value

      if (!code) {
        setFormError("Looks like the url has expired. Please try again.")
      } else if (password !== passwordConfirmation) {
        setFormError("Passwords do not match")
      } else {
        resetPassword({
          variables: {
            code,
            password,
            passwordConfirmation
          }
        }).then(({ data: { resetPassword } = {}}) => {
          if (resetPassword?.user?.email) {
            router.replace('/api/auth/signIn')
          }
        })
      }
    }}>
      <Box padding="0" gap="md" className="max-w-[300px]">
        <InputText fullWidth id="password" name="password" password={!showPassword} placeholder="New Password" />
        <InputText fullWidth id="password" name="passwordConfirmation" password={!showPassword} placeholder="Confirm New Password" />
        <Checkbox onChange={e => setShowPassword(e?.target?.checked)} label="Show password" /> 
        <Button type="submit" label={"Reset Password"} primary/> 
        {formError && <Paragraph textColor="hyper-red" darkTextColor="hyper-red" className="italic text-center" size="sm">{formError}</Paragraph>}
      </Box>
    </form>
  )
}