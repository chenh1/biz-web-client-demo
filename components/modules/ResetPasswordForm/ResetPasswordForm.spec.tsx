import { ResetPasswordForm } from "./ResetPasswordForm";
import { fireEvent, render, waitFor } from '@testing-library/react';
import { JSDOM } from "jsdom"
// import { resetPasswordMutation } from "../../../graphql/mutations/resetPassword";
const dom = new JSDOM()
global.document = dom.window.document
global.window = dom.window

jest.mock("next/navigation", () => ({
  useRouter() {
    return {
      prefetch: () => null,
      replace: (arg) => null
    };
  },
  useSearchParams: () => {
    return {
      get: () => "123"
    }
  }
}));

jest.mock("@apollo/client", () => ({
  useMutation: jest.fn(),
  gql: jest.fn()
}));

import { useMutation } from "@apollo/client";
const mockedUseMutation = useMutation as jest.Mock;

mockedUseMutation.mockImplementation((mutationDocument, onEvents) => {
  return [
    () => null,
    { data: {} }
  ]
})


it('renders Reset Password Form module', async () => {
  const { getByTestId, asFragment } = render(
    <ResetPasswordForm />
  )

  expect(asFragment()).toMatchSnapshot();
});
