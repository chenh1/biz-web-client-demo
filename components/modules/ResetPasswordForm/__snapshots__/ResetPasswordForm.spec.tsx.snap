// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Reset Password Form module 1`] = `
<DocumentFragment>
  <form>
    <div
      class=" grid  gap-8  p-2 max-w-[300px]"
    >
      <div
        class="relative "
      >
        <label
          class="  pointer-events-none font-body duration-300 absolute block text-md font-light text-night-black dark:text-max-white top-0 left-0  "
          for="password"
        >
          New Password
        </label>
        <input
          class="  max-w-full w-full text-night-black dark:text-max-white font-body bg-transparent border-b-2 focus:outline-none border-night-black dark:border-max-white"
          name="password"
          type="password"
        />
      </div>
      <div
        class="relative "
      >
        <label
          class="  pointer-events-none font-body duration-300 absolute block text-md font-light text-night-black dark:text-max-white top-0 left-0  "
          for="passwordConfirmation"
        >
          Confirm New Password
        </label>
        <input
          class="  max-w-full w-full text-night-black dark:text-max-white font-body bg-transparent border-b-2 focus:outline-none border-night-black dark:border-max-white"
          name="passwordConfirmation"
          type="password"
        />
      </div>
      <span
        class="relative"
      >
        <input
          class="  h-0 w-0 absolute"
          data-testid="checkbox"
          id="tempId"
          type="checkbox"
          value=""
        />
        <label
          class="  font-body relative flex my-[.6em] items-center text-night-black dark:text-max-white transition-colors duration-[250ms] ease-[cubic-bezier(.4,.0,.23,1)]"
          for="tempId"
        >
          <span
            class="pointer-events-none flex justify-center items-center mr-[.9em] w-[1.1em] h-[1.1em] bg-transparent border-2 border-solid rounded-sm cursor-pointer transition-all duration-[250ms] ease-[cubic-bezier(.4,.0,.23,1)] border-night-black dark:border-max-white"
          />
          <span
            class="pointer-events-none duration-500 opacity-1"
          >
            Show password
          </span>
          <ins
            class="pointer-events-none absolute block bottom-0 left-[2em] h-0 w-full overflow-hidden no-underline transition-[height] duration-300 ease-[cubic-bezier(.4,.0,.23,1)] "
          >
            <i
              class="pointer-events-none absolute bottom-0 not-italic text-hyper-red"
            >
              Show password
            </i>
          </ins>
        </label>
      </span>
      <button
        class="   w-full sm:w-auto cursor-pointer border-2 font-bold leading-none inline-block rounded-lg text-center font-headings relative text-max-white px-5 py-2.5   bg-hyper-red border-hyper-red dark:bg-hyper-red dark:border-hyper-red"
        data-testid="button"
        type="submit"
      >
        <div
          class="bg-hyper-red border-hyper-red dark:bg-hyper-red dark:border-hyper-red rounded-lg absolute left-0 top-0 w-full h-full "
        />
        <span
          class="pointer-events-none relative z-10"
        >
          Reset Password
        </span>
      </button>
    </div>
  </form>
</DocumentFragment>
`;
